### PR TITLE
Always complete the promise

### PR DESCRIPTION
**Pull request by takirala**
- Promise is used only in `ZkRepositoryList`
- A constant field is created in zookeeper class for the duration of wait.
- Helper method throws or returns the completed promise.

### CODE REVIEW

**Comment by takirala** [comment]
This the process I followed to test this : 
 - Spun up a local zookeeper on 2181 port. When cosmos boots up, it connects to this zookeeper.
 - Spun up a cluster in ccm. 
 - Run cosmos in debug mode locally with above dcos cluster. 
 - use dcos cli locally to search for a package
    - Dcos cli gets response for the search result.
    - Shutdown zookeeper and after one minute (when the cache invalidates itself) the search result is an error describing the timeout.

File: cosmos-common/src/main/scala/com/mesosphere/cosmos/error/TimeoutError.scala [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+
+import io.circe.Encoder
+import io.circe.JsonObject
+import io.circe.generic.semiauto.deriveEncoder
+
+final case class TimeoutError(
+  operation: String,
+  destination: String,
+  timeout: String
```
Comment (jsancio) [inline cosmos-common/src/main/scala/com/mesosphere/cosmos/error/TimeoutError.scala]: Make the type for `timeout` a `Duration` and apply the formatting in `def message`.

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/repository/ZkRepositoryList.scala [hunk 2]
```diff
@@ -186,20 +188,30 @@ final class ZkRepositoryList private (
...
-    implicit statsReceiver: StatsReceiver
+    implicit statsReceiver: StatsReceiver, timer: Timer
   ): ZkRepositoryList = {
     val repoList = new ZkRepositoryList(zkClient)
     repoList.start()
     repoList
   }
 
-  private val PackageRepositoriesPath: String = "/package/repositories"
+  private val PackageRepositoriesPath: String = "/cosmos/package/repositories"
```
Comment (jsancio) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/repository/ZkRepositoryList.scala]: The root path is not always `/cosmos`. Instead it is configured through the `ZooKeeperUri` `GlobalFlag`.

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/zookeeper/Clients.scala [hunk 3]
```diff
@@ -56,20 +55,21 @@ object Clients {
...
   private[this] def updateAcls(
+    zkUri: ZooKeeperUri,
     zkClient: CuratorFramework,
     aclProvider: ACLProvider
   ): Unit = {
     updateAcls(
       zkClient,
       aclProvider,
-      zkClient.getChildren.forPath("/").asScala.toList.map("/" + _)
+      zkClient.getChildren.forPath(zkUri.path).asScala.toList.map(zkUri.path + "/" + _ )
```
Comment (jsancio) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/zookeeper/Clients.scala]: :+1: 

See my comment in `ZkRepositoryList` but we need to also pass this URI to that class.

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/repository/ZkRepositoryList.scala [hunk 4]
```diff
@@ -158,7 +162,7 @@ final class ZkRepositoryList private (
     zkClient.getData().inBackground(
       new ReadHandler(promise)
     ).forPath(
-      ZkRepositoryList.PackageRepositoriesPath
+      zkUri.path + ZkRepositoryList.PackageRepositoriesPath
```
Comment (jsancio) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/repository/ZkRepositoryList.scala]: Let's make `PackageRepositoriesPath` `private[this]` and add a method that does this computation.

**Comment by takirala** [comment]
@jsancio Updated. is this what you had in mind?

**Pull request merged**